contracts/Linea_SpokePool.sol

```diff
@@ -0,0 +1,208 @@
+        _setL2MessageService(_l2MessageService);
+    }
+
+    /**
+     * @notice Change L2 USDC bridge address. Callable only by admin.
+     * @param _l2UsdcBridge New address of L2 USDC bridge.
+     */
+    function setL2UsdcBridge(IUSDCBridge _l2UsdcBridge) public onlyAdmin nonReentrant {
+        _setL2UsdcBridge(_l2UsdcBridge);
+    }
+
+    /**************************************
+     *        INTERNAL FUNCTIONS          *
+     **************************************/
+
+    /**
+     * @notice Wraps any ETH into WETH before executing base function. This is necessary because SpokePool receives
+     * ETH over the canonical token bridge instead of WETH.
+     */
+    function _preExecuteLeafHook(address l2TokenAddress) internal override {
+        if (l2TokenAddress == address(wrappedNativeToken)) _depositEthToWeth();
+    }
+
+    // Wrap any ETH owned by this contract so we can send expected L2 token to recipient. This is necessary because
+    // this SpokePool will receive ETH from the canonical token bridge instead of WETH. This may not be neccessary
+    // if ETH on Linea is treated as ETH and the fallback() function is triggered when this contract receives
+    // ETH. We will have to test this but this function for now allows the contract to safely convert all of its
+    // held ETH into WETH at the cost of higher gas costs.
+    function _depositEthToWeth() internal {
+        //slither-disable-next-line arbitrary-send-eth
+        if (address(this).balance > 0) wrappedNativeToken.deposit{ value: address(this).balance }();
+    }
+
+    function _bridgeTokensToHubPool(uint256 amountToReturn, address l2TokenAddress) internal override {
+        // SpokePool is expected to receive ETH from the L1 HubPool, then we need to first unwrap it to ETH and then
+        // send ETH directly via the Canonical Message Service.
+        if (l2TokenAddress == address(wrappedNativeToken)) {
+            WETH9Interface(l2TokenAddress).withdraw(amountToReturn); // Unwrap into ETH.
+            IMessageService(l2MessageService).sendMessage{ value: amountToReturn }(hubPool, 0, "");
+        }
+        // If the l1Token is USDC, then we need sent it via the USDC Bridge.
+        else if (l2TokenAddress == l2UsdcBridge.usdc()) {
+            IERC20(l2TokenAddress).safeIncreaseAllowance(address(l2UsdcBridge), amountToReturn);
+            l2UsdcBridge.depositTo(amountToReturn, hubPool);
+        }
+        // For other tokens, we can use the Canonical Token Bridge.
+        else {
+            IERC20(l2TokenAddress).safeIncreaseAllowance(address(l2TokenBridge), amountToReturn);
+            l2TokenBridge.bridgeToken(l2TokenAddress, amountToReturn, hubPool);
+        }
+        emit LineaTokensBridged(l2TokenAddress, hubPool, amountToReturn);
+    }
+
+    function _requireAdminSender() internal view override {
+        require(IMessageService(l2MessageService).sender() == crossDomainAdmin, "ONLY_COUNTERPART_GATEWAY");
```

Comment:
Have you tested that no one else can call a function protected by this internal func?

Reply:
crossDomainAdmin is set to the HubPool address, right?

Reply:
> Have you tested that no one else can call a function protected by this internal func?

Yes, I tried to call a few functions with an EOA for example.

> crossDomainAdmin is set to the HubPool address, right?

Yes, it is set to the address of the HubPool

Reply:
> Yes, I tried to call a few functions with an EOA for example.

Did you also test calling over the L1->L2 bridge from an address that isn't the HubPool?

Reply:
> Did you also test calling over the L1->L2 bridge from an address that isn't the HubPool?

yes I also tested that